express-notes-api/server.js

```diff
@@ -0,0 +1,148 @@
+import express from 'express';
+import { writeFile } from 'node:fs/promises';
+import { readFileSync } from 'node:fs';
+
+// --- Reusable functions ---
+function handleError(res, err) {
+  console.error('/api/notes error', err);
+  res.status(500).send({ error: 'An unexpected error occurred.' });
+}
+
+function evaluateId(res, journal, id) {
+  const validIds = Object.keys(journal.notes);
+  const errorMsg = `${id} is not a valid ID. Please provide a valid ID: ${validIds.join(', ')}`;
+  if (isNaN(id) || Number(id) < 0) {
+    res.status(400).send(errorMsg);
+  } else if (!validIds.includes(id)) {
+    res.status(404).send(errorMsg);
+  } else {
+    return true;
+  }
+}
+
+function evaluateReqBody(req, res, action) {
+  let userInputs = Object.keys(req.body);
+  userInputs = userInputs.filter((key) => key !== 'id');
+  switch (action) {
+    case 'delete':
+      if (userInputs.length > 0) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised the above inputs were not be processed. Please try again.`);
+      } else {
+        return true;
+      }
+      break;
+    case 'put':
+    case 'post':
+      if (!userInputs.includes('content') || userInputs.length > 1) {
+        res.status(400).send(`Input: ${JSON.stringify(req.body)}
+  Be advised only notes assigned to "content" will be processed. Please try again.`);
+      } else {
+        return true;
+      }
+  }
+}
+
+/** Takes an Express response and user-provided id as inputs.
+ *  If there is an ID, return the entry for that given ID. Otherwise, return all entries. */
+async function loadNotes(res, id) {
+  try {
+    const data = await readFileSync('data.json', 'utf8');
```

Comment:
Never synchronously read files (`readFileSync`) when responding to web requests: it will prevent the server from responding to multiple requests at the same time.